### ISSUE TITLE
Investigate if OpenSSL (libcrypto) vendored in here can be FIPS certified

### ISSUE THREAD

**Issue by xnox** [rhboot/shim#586]
Life would be easier if libcrypto inside shim was FIPS certified, as for example it would then not be needed for everyone to ship .hmac files for vmlinuz.

Looking at the code it seems like we currently reverted to using 1.0.2k OpenSSL version. Ubuntu actually has a FIPS certified 1.0.2g OpenSSL thus it would be interesting to see if this can be achieved.

I am also slightly concerned by how old our OpenSSL is, even if it is unlikely that any bugs or CVEs affect our reduced codebase.

In https://github.com/rhboot/shim/commit/1d39ada8cb336d9e7c156be7526b674851fbdd40 @vathpela reverted updates to OpenSSL due to incompatibilities with existing certs => have the certs that broke got stored somewhere as test cases? I separately ponder if we can or should work with OpenSSL upstream to contribute "UEFI" mode to OpenSSL to accept more certs.

Basically a can of worms =)

**Comment by steve-mcintyre** [comment]
Is it worth trying to use a really old openssl for FIPS certification any more? 140-3 with openssl 3 looks like a much better option, surely?

**Comment by xnox** [comment]
> Is it worth trying to use a really old openssl for FIPS certification any more? 140-3 with openssl 3 looks like a much better option, surely?

Yes, if it works... Note upgrade to 1.1.0 got reverted due to production secureboot certs being rejected by new openssl thus failing to boot things validly signed.

And it seems nobody is looking to upgrade the vendored openssl and/or fixup openssl to still work.

**Comment by jsetje** [comment]
We probably need to figure out how to get to a newer openssl regardless. The notes in https://github.com/rhboot/shim/commit/1d39ada8cb336d9e7c156be7526b674851fbdd40 also make me thins that we should look into some compatibility. 

Let me make sure that I understand what you're asking for WRT FIPS: Are you hoping for a FIPS certified shim binary? I'm not sure that is worth the effort if it's even possible. Or do you just want to unlock CERTs that a FIPS compliant openssl is happy with for use in shim? I do agree that the latter seems rather desirable.

**Comment by xnox** [comment]
Pondering if https://github.com/smuellerDD/leancrypto/issues/9 is a good choice for replacing cryptography in shim. That library targets bare-metal / freestanding code. has been fips certified before. and has PQC.

**Comment by smuellerDD** [comment]
Now that leancrypto has full support for PKCS7 and X.509 as well as a native compile target for EFI, I would be wondering whether I should give it a try to add it to shim. This would mean the use of a crypto lib that can easily be kept up to date without any porting effort for the EFI environment. See https://leancrypto.org/leancrypto/build_instructions/efi/index.html

The leancrypto library als allows a fine grained compile-time configuration to only select the needed algorithms.

Note that also accelerated algo implementations are available in the EFI environment.

**Comment by smuellerDD** [comment]
Just as a note - I am in the process of adding support for leancrypto to shim. The code is roughly there, but needs some testing and review before releasing. The changes keep the OpenSSL for RSA-based certificates, but uses leancrypto for everything else (accelerated SHA, random number generation, PQC, X.509, PKCS7). This allows PQC (ML-DSA and SLH-DSA) signatures as well as composite signatures of classic + PQC (ML-DSA + ED448 or ED25519).

**Comment by smuellerDD** [comment]
Since I am not fully sure about the process MS applies, are you saying that 
RSA is not needed for this chaining of the shim certificates to the MS root?

The core question therefore is: can I completely remove OpenSSL already now? 
Leancrypto does not offer the support for RSA and SHA-1 and I do not plan do 
add it. Thus, when you say we do not need RSA at all, I could remove all 
OpenSSL code. Otherwise, I would need to leave it and invoke the RSA support, 
for example, when leancrypto cannot parse the processed certificate (e.g. when 
RSA is there).

**Comment by smuellerDD** [comment]
A preview of the code can now be found at https://github.com/smuellerDD/shim/tree/leancrypto

Please note that this code is currently work in progress. Specifically it has only been compile-tested so far. I need to start testing it (although I first have to see how to test a boot loader).

Yet, any comments are welcome.

**Comment by smuellerDD** [comment]
To simplify the code review https://github.com/smuellerDD/shim/tree/leancrypto2 is added which avoids moving around existing code and simply adding the leancrypto parts.

**Comment by smuellerDD** [comment]
PQC and FIPS 140 Support for Shim Boot Loader

Hello,

I have completed the development to add PQC support to the shim boot loader by linking it with the leancrypto library. In addition, the build scripts are updated to properly build leancrypto alongside shim and link them to the final EFI executable. Also, the certificate generation and signing of the binary to be booted is added and works. With the build system update, the binary is created that is intended to be used for testing by using the command `mkosi --firmware-variables mkosi/mkosi.output/ovmf_vars.fd --distribution centos build`.

However, the following issue currently prevents me from booting the image and performing testing:

```
‣  Running post-output script /home/sm/tmp/shim.leancrypto/mkosi/mkosi.postoutput…
No OVMF vars template found for x86-64
‣ "/work/postoutput" returned non-zero exit code 1.
```

The mkosi.postoutput script looks for some OVMF.fd file which is not installed with the centos template (or other templates). I am also not finding guidance in the readmes where to get it. Further, when looking in other systems like OpenSUSE Tumbleweed, I find a sizeable number of OVMF.fd files.

Thus, may I ask where I can get this file from or which file should I apply from other distros? Or perhaps may I ask whether there is a more appropriate place where I can state that question?

PS: In case somebody wants to reproduce the current state: see [1], specifically the section "Quick Start". For this, use the GIT Head of leancrypto found in [2]. Note, as currently the sbsigntools do not support generation of signatures using PQC, leancrypto implements drop-in replacements which are applied with the ukify tool in mkosi.finalize.

PPS: The leancrypto library is build as a separate static library that is linked into shim during build time. This implies that no artificial copy of a crypto library needs to be integrated into shim (which is currently the case with the ancient OpenSSL 1.0.2 copy). Instead, a separately maintained and current crypto library would be available to be used with shim. Furthermore, with leancrypto accelerated implementations of PQC but also of other cryptographic algorithms are at shim's disposal. Of course, one can prepare a static EFI binary libleancrypto.a that can be linked with shim time and again for providing a stable copy for the MS signing process.

PPPS: With the leancrypto library, a FIPS-140 compliant solution for secure boot would be available as discussed in [3].

[1] https://github.com/smuellerDD/shim/blob/leancrypto2/README.leancrypto.md

[2] https://github.com/smuellerDD/leancrypto

[3] https://github.com/smuellerDD/leancrypto/blob/master/README.fips.md

Thanks a lot
Stephan

**Comment by julian-klode** [comment]
Thanks Stephan!

While I can't comment on crypto library choices, not being a maintainer of shim, just one of the downstream distros, I have a couple notes:

1. It does not support ECDSA with the NIST P Curves which may play a significant role in this space
2. It links against gnu-efi whereas shim is sort of trying to get rid of it (stuck on a fork of it with more changes)
3. RSA is going to stick around for the next decade or so, it's a bit wasteful to have both leancrypto and old OpenSSL code paths.

I'd very much like if we could have the whole X.509, PE, Authenticode parsing bits handled in Rust and only deal with crypto primitives in an unsafe language - that would significantly lower the attack surface.

Ultimately regardless of possible implementation avenues, PQC support relies on Microsoft to publish updated Authenticode; it's not possible yet to actually use PQC.

**Comment by smuellerDD** [comment]
Regarding ECDSA: I am working with CMVP and BSI together. Albeit they prefer composite signatures, the choice of the "classic" signature type is usually secondary to them provided that the PQC algorithm is approved (see BSI TR02102 or NIST IR 8547).

Regarding gnu-efi: This is only needed for the memory allocation functions of AllocatePool and AllocatePages and their respective free functions (see https://github.com/smuellerDD/leancrypto/blob/master/internal/src/memory_support_efi.c). If shim wants to move away from gnu-efi, all that is needed is to link to the shim functions providing this logic - which should be straight forward given that leancrypto is intended to be used in different environments (e.g. it works equally well in the Linux kernel).

Regarding RSA: I agree that in general RSA may be around for time to come. Yet, leancrypto should not receive RSA as it is on its way out. Most folks today want ECC and move away from RSA. Thus, RSA is only legacy. That said, for the transition, keeping the old accepted and reviewed RSA code should do no harm. Yes, it will enlarge the binary a bit as there are two X.509 parsers in it, but I think keeping the old tested code trumps any re-implementation of RSA in leancrypto.

Regarding Rust: leancrypto has Rust bindings and should fare well when a Rust implementation is used. And note, having an implementation of C code that can be tested, executed and analyzed independently of the confinements of shim should be a great benefit - this would be offered by leancrypto.

Regarding MS: I totally agree, albeit the format MS defines allows arbitrary signature sizes and thus should allow PQC algos without a format change (see an implementation of the Authenticode in https://github.com/smuellerDD/leancrypto/blob/master/apps/src/authenticode_SpcIndirectDataContent_asn1 as used by https://github.com/smuellerDD/leancrypto/blob/master/apps/src/sbsign.c). Getting the code in place in general and support the booting of PQC-signed boot loaders and kernels should and could already be done today. The final step of having the shim boot loader itself signed by PQC relies on PQC support in the UEFI layer.
